Pausable node upgrade

starbops opened the pull request:
<!-- 
!IMPORTANT!
Please do not create a Pull Request without creating an issue first.
-->

#### Problem:
<!-- Explain the problem you aim to resolve in this PR. -->

In the current Harvester Upgrade design, it's impossible to pause a node entering the node-upgrade state beforehand. This causes headaches, such as more extended downtime for user workloads.

#### Solution:
<!-- Example: When "Adding a function to do X", explain why it is necessary to have a way to do X. -->

Let users specify which nodes should be paused before entering the node-upgrade state —specifically, before the pre-draining state, by configuring the newly introduced `nodeUpgradeOption` under the `upgrade-config` setting.

Such a setting will be translated to a node-to-action map in an annotation and attached to the Upgrade CR by the mutator upon creation. This intermediate representation is the source of truth that the controllers honor. An example is like below:

```yaml
...
metadata:
  harvesterhci.io/node-upgrade-pause-map: '{"bravo-1-tink-system": "pause", "bravo-2-tink-system": "pause"}'
  ...
```

When the upgrade proceeds to the node-upgrade phase, the embedded Rancher will pick a node to conduct the node-upgrade tasks. If such a node is marked as `pause` in the Upgrade CR's annotation, the pre-drain job won't be created by the Harvester upgrade controller; otherwise, the job is created as usual. When any node is being paused, the per-node upgrade status field will show the pause state. For instance, the following indicates that node `bravo-1-tink-system` is paused for node upgrade:

```yaml
status:
  nodeStatuses:
    bravo-1-tink-system:
      state: Node-upgrade paused
      reason: AdministrativelyPaused
      message: Node upgrade paused as requested by the user
    bravo-2-tink-system:
      state: Images preloaded
```

To un-pause a paused node, annotate the corresponding annotation in the Upgrade CR to `unpause`. The upgrade process will proceed until the next request-to-pause node (if any).

```yaml
...
metadata:
  harvesterhci.io/node-upgrade-pause-map: '{"bravo-1-tink-system": "unpause", "bravo-2-tink-system": "pause"}'
  ...
```

> [!NOTE]
> It's hard to know which node will be upgraded next from Harvester's viewpoint (the embedded Rancher decides the order). Harvester, as a consumer, can at best only pause a node from undergoing an upgrade by not creating the pre-drain job. However, the node will still be put into the cordon state (but not drained yet, fortunately). In the end, users should still be able to move things out from that node.

> [!NOTE]
> For experienced users (or automation, probably), it's entirely feasible to add the pause node annotations directly to the Upgrade CR without configuring the `upgrade-config` setting. As long as the node hasn't ever entered the node-upgrade state, the annotation is deemed valid and will take effect. The `nodeUpgradeOption` in the `upgrade-config` setting provides a simple way for users to configure without having to annotate the Upgrade CR every time.

#### Related Issue(s):
<!--
Use `Issue #<issue number>` or `Issue harvester/harvester#<issue number>` or `Issue (paste link of issue)`. DON'T use `Fixes #<issue number>` or `Fixes (paste link of issue)`, as it will automatically close the linked issue when the PR is merged.
-->

#8980

#### Test plan:
<!-- Describe the test plan by steps. -->

**Single-node upgrade**

1. Prepare a single-node cluster provisioned using a custom-built ISO image containing the enhancement
2. Configure the `upgrade-config` setting, adding the following options:
   ```
   ...
   "nodeUpgradeOption": {
     "strategy": {
       "mode": "manual"
     }
   },
   ...
   ```
3. Trigger an upgrade with the same ISO image
4. Observe the created Upgrade CR, it should contain the rendered annotation `harvesterhci.io/<node_name>: pause`
5. The upgrade should be paused at the node-upgrade phase until the above annotation is updated to `unpause`
   - The `NodesUpgraded` condition should remain `Unknown`
   - The `nodeStatuses` for the node should becomes `AdministrativelyPaused`
6. Unpause the node by updating the Upgrade CR's annotation for the node to `unpause`
7. The upgrade should immediately continue
8. The upgrade should complete successfully

**Multi-node upgrade**

1. Prepare a 3-node cluster provisioned using a custom-built ISO image containing the enhancement
2. Configure the `upgrade-config` setting, adding the following options:
   ```
   ...
   "nodeUpgradeOption": {
     "strategy": {
       "mode": "manual",
       "pauseNodes": [
         "<node-0>",
         "<node-2>"
       ]
     }
   },
   ...
   ```
   Only two of the three nodes specified in the `pauseNodes` list will be paused during node upgrades.
3. Trigger an upgrade with the same ISO image
4. Observe the created Upgrade CR, it should contain the rendered annotations `harvesterhci.io/<node_name>: pause`
5. The upgrade should be paused when one of the above two nodes entering the node-upgrade state, until the annotation is updated to `unpause`
   - The `NodesUpgraded` condition should remain `Unknown`
   - The `nodeStatuses` for the node should becomes `AdministrativelyPaused`
6. Unpause the node by updating the Upgrade CR's annotation for the node to `unpause`
7. The upgrade should immediately continue
8. Repeat step 5 (if all the pause nodes annotations are updated to unpause, go to step 9)
9. The upgrade should complete successfully

#### Additional documentation or context

starbops committed:
feat(setting): introduce nodeUpgradeOption fields to upgrade-config setting

The commit defines:

- basic structures
- default values

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
test(setting): add validation on nodeUpgradeOption

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
feat(upgrade): add upgrade mutator

When creating upgrade resources, node-upgrade related configs will be
transformed and added to them. Specifically, the pause node upgrade
annotations.

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
feat(upgrade): support pausable node upgrade

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
fix: register upgrade mutator

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
fix: pass on missing nodes upon unpause events

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
feat(upgrade): reflect node upgrade pause on condition

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
refactor: address golangci-lint raised issues

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
fix(upgrade): trigger machine-plan secret reconcil correctly

Append the "-machine-plan" suffix to the machine name to correctly
rending the machine-plan secret name.
Also, pausable node upgrade logic for single-node upgrade should not
interfere with multi-node upgrade flow.

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

brandboat commented on pkg/controller/master/upgrade/secret_controller.go:
```diff
@@ -75,9 +76,17 @@ func (h *secretHandler) OnChanged(_ string, secret *v1.Secret) (*v1.Secret, erro
 		if secret.Annotations[rke2PreDrainAnnotation] != secret.Annotations[preDrainAnnotation] {
+			if shouldPauseNodeUpgrade(upgrade, nodeName) {
+				logrus.Infof("Pause creating pre-drain job on %s", nodeName)
+				setPausedCondition(upgradeCpy, v1.ConditionTrue, "NodeUpgrade", fmt.Sprintf("node upgrade for %s is administratively paused as requested", nodeName))
```
I was just thinking — do we have any plan to support upgrading multiple nodes at the same time in the future?
If yes, maybe it would make more sense to put the pause status under .status.nodeStatuses[node_name], since the current condition might get overwritten when another node gets paused.

But for now, since we only upgrade nodes one by one, I’m totally fine with the current approach.

brandboat commented on pkg/controller/master/upgrade/common.go:
```diff
@@ -53,6 +53,18 @@ fi
+	if val != "pause" {
+		return false
+	}
+	return true
```
nit: 
```suggestion
    return val == "pause"
```

brandboat reviewed (commented):
Thanks for the patch!

ihcsim commented on pkg/controller/master/upgrade/register.go:
```diff
@@ -21,6 +28,43 @@ const (
+			if val == "pause" {
+				continue
+			}
+			nodeName, ok := extractNodeName(key)
```
This means we can only unpause one node at a time, right?

ihcsim commented on pkg/controller/master/upgrade/secret_controller.go:
```diff
@@ -75,9 +76,17 @@ func (h *secretHandler) OnChanged(_ string, secret *v1.Secret) (*v1.Secret, erro
+				setPausedCondition(upgradeCpy, v1.ConditionTrue, "NodeUpgrade", fmt.Sprintf("node upgrade for %s is administratively paused as requested", nodeName))
+				break
+			}
+			setPausedCondition(upgradeCpy, v1.ConditionFalse, "", "")
```
can we put the pre-drain job in `suspend` mode, instead of `break`-ing at line 87? it will make future support of multi-nodes upgrade easier.

ihcsim replied:
Sorry for being unclear earlier, what i meant was create the job anyway, but in [`suspend` mode](https://kubernetes.io/docs/concepts/workloads/controllers/job/#suspending-a-job) so the pod won't run, until the node is unpaused. What you have works too - thanks.

starbops committed:
fix(upgrade): dup detect for pause nodes option

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
feat(upgrade): reflect pause in per-node upgrade state

Rename the "Paused" condition to "Degraded" to better support the
multi-node upgrade cases in the future.

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

brandboat reviewed (dismissed):
LGTM, thanks!

w13915984028 commented on pkg/webhook/resources/upgrade/mutator.go:
```diff
@@ -0,0 +1,132 @@
+
+	for _, nodeName := range nodeNames {
+		patchOps = append(patchOps,
+			fmt.Sprintf(`{"op": "add", "path": "/metadata/annotations/harvesterhci.io~1%s", "value": "pause"}`, nodeName))
```
the annotation key has a name field length limit 63 chars, need to double check if `nodeName` can exceed this limit

w13915984028 commented on pkg/controller/master/upgrade/common.go:
```diff
@@ -53,6 +53,15 @@ fi
 )
 
+func shouldPauseNodeUpgrade(upgrade *harvesterv1.Upgrade, nodeName string) bool {
+	key := fmt.Sprintf("%s/%s", util.Prefix, nodeName)
```
replace `util.Prefix` with `util.GetHarvesterPrefix()`

w13915984028 commented on pkg/controller/master/upgrade/register.go:
```diff
@@ -21,6 +29,43 @@ const (
+func (h *handler) NotifyUnpausedMachinePlanSecret(_ string, _ string, obj runtime.Object) ([]relatedresource.Key, error) {
+	if upgrade, ok := obj.(*harvesterv1.Upgrade); ok {
+		for key, val := range upgrade.Annotations {
+			if val == "pause" {
```
could `pause` be defined as a const?

w13915984028 reviewed (commented):
Some questions, thanks.

starbops committed:
fix(upgrade): change to node upgrade pause map

Refactor the code to use a pause map annotation instead of annotations
for individual nodes. This prevents the annotation key length limit
being hit.

Pause nodes will be paused upon node upgrades. Update the value to
unpause will unblock the node upgrade. In a later reconciliation, the
controller will remove the already unpaused node from the pause map
annotation.

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

ihcsim commented on pkg/webhook/resources/upgrade/mutator.go:
```diff
@@ -0,0 +1,143 @@
+		},
+	}
+}
+func (m *upgradeMutator) Create(_ *types.Request, newObj runtime.Object) (types.PatchOps, error) {
```
Since we are watching only `create` events, susbsequent additions to the upgrade annotations will have no effects, right?

ihcsim commented on pkg/settings/settings_helper.go:
```diff
@@ -168,13 +166,38 @@ type ImagePreloadOption struct {
+	// PauseNodes contains the node names that should be paused before entering the pre-draining state during the node
+	// upgrade phase. It is only honored when the Mode is "manual". If the Mode is "manual" but no node names are
+	// provided, all nodes will be paused before entering the pre-draining state.
+	PauseNodes []string `json:"pauseNodes,omitempty"`
```
what do you think of using node labels to allow more flexibility with grouping and selections? for per node identification, user can also use the `kubernetes.io/hostname` label.

ihcsim commented on pkg/webhook/resources/setting/validator.go:
```diff
@@ -1814,7 +1814,61 @@ func validateUpgradeConfigHelper(setting *v1beta1.Setting) (*settings.UpgradeCon
 
-func validateUpgradeConfigFields(setting *v1beta1.Setting) error {
+func checkNodesDuplicates(nodeNames []string) error {
+	duplicates := ds.SliceFindDuplicates(nodeNames)
```
why not just dedupe it with `ds.SliceDedupe`?

starbops replied:
The returned error message clearly shows what is duplicated, which provides a better user experience.

starbops commented on pkg/controller/master/upgrade/common.go:
```diff
@@ -53,6 +54,59 @@ fi
+	value, ok := upgrade.Annotations[util.AnnotationNodeUpgradePauseMap]
+	if !ok {
+		return nil, nil
+	}
```
The annotation map stores the desired action for specific nodes upon node upgrade. The actual state of the nodes are recorded in the `status.nodeStatuses`. The secret controller will factor in the `status.nodeStatuses` (current node state), the secret annotations, and the annotation map (desired node upgrade action). For instance, if the current node state is "image preloaded", the secret has "pre-drain" annotated, and the map says the node should be paused, the node upgrade for the node will be paused. Another example, if the current node state is "node-upgrade paused", the secret has "pre-drain" annotated, and the map says the node should be unpaused, the node upgrade for the node will be paused.

starbops committed:
fix(upgrade): re-enqueue secret to proceed with upgrades

Unpausing a node for node upgrade consists of two steps:
1. Signal the secret controller to remove the entry from the pause map
   annotation and re-enqueue the secret object
2. Secret controller checks the pause map again and continues the
   upgrade

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
fix(upgrade): drop degraded condition

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

w13915984028 commented on pkg/webhook/resources/upgrade/mutator.go:
```diff
@@ -0,0 +1,143 @@
+		patchOps = append(patchOps, `{"op": "add", "path": "/metadata/annotations", "value": {}}`)
+	}
+
+	nodeMap := make(map[string]string, len(nodeNames))
```
add below before L125, as the passed nodeNames could have zero valid elements
```
if len(nodeNames) == 0 {
	return nil
}
```

starbops committed:
fix(upgrade): ignore non-existent nodes in pause node list

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
fix(upgrade): 2-step single node upgrade pause handling

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
feat(upgrade): validate pause map annotation

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

starbops committed:
feat(upgrade): leave unpause record in the pause map

Unpaused node records are kept in the pause map annotation for
references. The machine-plan secret enqueue criteria is adjust as a
result. Now it will ensure only those machine-plan secrets whose nodes
are in paused state and are requested to be unpaused are enqueued.

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

w13915984028 reviewed (dismissed):
LGTM, thanks.

starbops committed:
refactor: reuse getNodeUpgradePauseMap to load map

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

brandboat commented on pkg/controller/master/upgrade/secret_controller.go:
```diff
@@ -75,9 +79,18 @@ func (h *secretHandler) OnChanged(_ string, secret *v1.Secret) (*v1.Secret, erro
+
 	switch upgrade.Status.NodeStatuses[nodeName].State {
 	case nodeStateImagesPreloaded:
 		if secret.Annotations[rke2PreDrainAnnotation] != secret.Annotations[preDrainAnnotation] {
```
gentle ping @starbops, could you please fix this? Thank you!

`secret.Annotations[rke2PreDrainAnnotation] != secret.Annotations[preDrainAnnotation]` ->
```
!isPreDrained(secret)
```

starbops committed:
feat(upgrade): skip mutating upgrade if pause map anno is specified

The upgrade mutator skips translating and adding the pause map
annotation from the setting if such an annotation is already
explictly specified in the upgrade cr upon creation.

Signed-off-by: Zespre Chang <zespre.chang@suse.com>

w13915984028 reviewed (approved):
LGTM, thanks.

ihcsim commented on pkg/controller/master/upgrade/upgrade_controller.go:
```diff
@@ -397,6 +397,22 @@ func (h *upgradeHandler) OnChanged(_ string, upgrade *harvesterv1.Upgrade) (*har
+				logrus.Infof("Pause node-upgrade job creation for node %s", singleNodeName)
+				setNodeUpgradeStatus(toUpdate, singleNodeName, nodeStateUpgradePaused, "AdministrativelyPaused", "Node upgrade paused as requested by the user")
+				return h.upgradeClient.Update(toUpdate)
+			}
```
It's a TIOLI optimization to reduce duplicate code with the `shouldPauseNodeUpgrade()` 😄. Yeah, I did notice it's trying to follow the same pattern as the `switch` in the secret controller. It's fine.

Vicente-Cheng reviewed (approved):
just a nit question, overall lgtm

starbops commented on pkg/controller/master/upgrade/secret_controller.go:
```diff
@@ -75,17 +78,43 @@ func (h *secretHandler) OnChanged(_ string, secret *v1.Secret) (*v1.Secret, erro
-			if err := h.createHookJob(upgrade, nodeName, upgradeJobTypePreDrain, nodeStatePreDraining); err != nil {
-				return nil, err
-			}
+		if isPreDrained(secret) {
```
Actually, the function only checks the machine-plan secret's annotations. In reality, there will be three possible combinations:

1. Both the `pre-drain` and `pre-hook` annotations are missing; this implies the node is not under upgrade.
2. Only the `pre-drain` annotation exists; this means Rancher notifies us (Harvester) to initiate the pre-drain job, and it is waiting for us to annotate the `pre-hook` annotation once the job is done.
3. Both the `pre-drain` and `pre-hook` annotations exist and have the same value; this represents that we have already notified Rancher to continue with the drain, which implies the pre-drain job has been completed.

The secret controller does not care whether the pre-/post-drain job is running or finished. It only focuses on the machine-plan secrets.

So, according to the above cases, the `isPreDrained` function should return:

1. `true`
2. `false`
3. `true`

The function name can be misleading because in case 1, it's actually "not-yet-pre-drained". Only cases 2 & 3 fit in the context of the function name. We can rename it to something like `isUnderPreDrain` and negate the condition.

starbops committed:
refactor: rename function to better describe the context

Signed-off-by: Zespre Chang <zespre.chang@suse.com>
